File: implementations/fastapi/examples/src/auth/application/command/sign_in_handler.py
from dataclasses import dataclass

from ...domain.errors import InvalidCredentialsError
from ...domain.repository import CredentialRepository
from ..service.auth_service import AuthService
from ..service.password_hasher import PasswordHasher


@dataclass
class SignInCommand:
    user_id: str
    password: str


class SignInHandler:
    def __init__(self, repo: CredentialRepository, password_hasher: PasswordHasher, auth_service: AuthService) -> None:
        self._repo = repo
        self._password_hasher = password_hasher
        self._auth_service = auth_service

    async def execute(self, cmd: SignInCommand) -> str:
        credentials, _ = await self._repo.find_credentials(page=0, take=1, user_id=cmd.user_id)
        credential = credentials[0] if credentials else None
        # Responds with the same error whether the username doesn't exist or the password doesn't match
        # — prevents user enumeration
        if credential is None:
            raise InvalidCredentialsError()

        is_valid = await self._password_hasher.verify(cmd.password, credential.password_hash)
        if not is_valid:
            raise InvalidCredentialsError()

        return self._auth_service.issue_token(credential.user_id)

File: implementations/go/examples/internal/application/command/sign_in_handler.go
package command

import (
	"context"
	"errors"

	"github.com/example/account-service/internal/domain/credential"
)

type SignInCommand struct {
	UserID   string
	Password string
}

type SignInHandler struct {
	repo           credential.Repository
	passwordHasher PasswordHasher
	tokenIssuer    TokenIssuer
}

func NewSignInHandler(repo credential.Repository, passwordHasher PasswordHasher, tokenIssuer TokenIssuer) *SignInHandler {
	return &SignInHandler{repo: repo, passwordHasher: passwordHasher, tokenIssuer: tokenIssuer}
}

// Handle issues a token only after verifying the password against the
// stored hash. It responds with the same error
// (credential.ErrInvalidCredentials) whether the user ID doesn't exist or
// the password is wrong — distinguishing the two would let an attacker
// guess which user IDs exist (user enumeration).
func (h *SignInHandler) Handle(ctx context.Context, cmd SignInCommand) (string, error) {
	c, err := credential.FindOne(ctx, h.repo, cmd.UserID)
	if err != nil {
		if errors.Is(err, credential.ErrNotFound) {
			return "", credential.ErrInvalidCredentials
		}
		return "", err
	}

	valid, err := h.passwordHasher.Verify(cmd.Password, c.PasswordHash)
	if err != nil {
		return "", err
	}
	if !valid {
		return "", credential.ErrInvalidCredentials
	}

	return h.tokenIssuer.Sign(c.UserID)
}
